### PR TITLE
Fps plugin

### PR DESCRIPTION
**Pull request by LeviSchuck**
This depends on PR #1349 
Only pay attention to what is in `runelite-client/src/main/java/net/runelite/client/plugins/fps/` or review https://github.com/runelite/runelite/pull/1350/commits/eb2f738891b253f62aac122dcdd4894910f4e79d

Add FPS plugin, main features: Show FPS, Limit FPS always, Limit FPS on unfocus

Main motivation: OSRS likes to burn my laptop's battery in the form of heat with its software rendering. I want my laptop to stay cool and not need a charging cord for hours.

On my desktop, this goes from 50% CPU to about 10-20% when around 15fps. It depends on whether or not the game is in expanded render mode.

fps-plugin

You may notice the display smooths down to about 16, this is the number given by client.getFPS, which is not quite accurate as far as counting the milliseconds goes and is averaged over time.

### CODE REVIEW

File: runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsConfig.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+	@ConfigItem(
+		keyName = "maxFps",
+		name = "FPS target",
+		description = "Desired max frames per second",
+		position = 3
+	)
+	default int maxFps()
+	{
+		return 15;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsConfig.java]: This should probably default to 50

File: runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsOverlay.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+	void onFocusChanged(FocusChanged event)
+	{
+		isFocused = event.isFocused();
+		changeColor();
+	}
+
+	void reloadConfig()
+	{
+		enabled = config.drawFps();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsOverlay.java]: Why is this necessary? There is no reason to not just use config.abc directly as you already have it in the hand

File: runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsOverlay.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+		alwaysEnabled = config.alwaysLimitFps();
+		unfocusedEnabled = config.unfocusedLimitFps();
+		changeColor();
+	}
+
+	private void changeColor()
+	{
+		if (alwaysEnabled || (unfocusedEnabled && !isFocused))
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsOverlay.java]: I think this colorization is a bit confusing, you should probably color it red when it's like far off from desired FPS target, and then maybe yellow if in between and white if at target

File: runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsOverlay.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+	@Override
+	public Dimension render(Graphics2D graphics)
+	{
+		if (enabled)
+		{
+			final int fps = client.getFPS();
+			if (fps < FPS_SIZE)
+			{
+				final int width = client.isResized()
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsOverlay.java]: Why have it special cased like this?

File: runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsPlugin.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+@PluginDescriptor(
+	name = "FPS Control",
+	enabledByDefault = false
+)
+public class FpsPlugin extends Plugin
+{
+	static final String CONFIG_GROUP_KEY = "fpscontrol";
+
+	@Provides
+	FpsConfig provideConfig(ConfigManager configManager)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsPlugin.java]: Move methods below field declarations

File: runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsOverlay.java (hunk 2 above)
Comment (LeviSchuck) [reply runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsOverlay.java]: cache-line performance. But for ease of access, I'll forgo those habits.

File: runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsOverlay.java (hunk 4 above)
Comment (LeviSchuck) [reply runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsOverlay.java]: Indeed it seems to be unnecessary. I think I saw somewhere else in the client do this and I thought it was because of performance concerns.

File: runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsConfig.java (hunk 1 above)
Comment (LeviSchuck) [reply runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsConfig.java]: Seems to be the majority vote from the discussion.

File: runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsDrawListener.java [hunk 6]
```diff
@@ -90,7 +86,7 @@ void onFocusChanged(FocusChanged event)
 	@Override
 	public void accept(BufferedImage bufferedImage)
 	{
-		if (alwaysEnabled || (unfocusedEnabled && !isFocused))
+		if (config.alwaysLimitFps() || (config.unfocusedLimitFps() && !isFocused))
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsDrawListener.java]: Invert this condition (to prevent having giant if blocks in size of the entire method)

File: runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsOverlay.java [hunk 7]
```diff
@@ -88,48 +81,33 @@ private FpsOverlay(FpsConfig config, Client client)
...
-		}
+	private Color getFpsValueColor() {
+		return isEnforced() ? Color.red : Color.yellow;
 	}
 
 	@Override
 	public Dimension render(Graphics2D graphics)
 	{
-		if (enabled)
+		if (config.drawFps())
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsOverlay.java]: See my previous comment

**Comment by LeviSchuck** [comment]
Rebased again again. Should be clean as it no longer includes the draw manager pr.

File: runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsConfig.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+)
+public interface FpsConfig extends Config
+{
+	@ConfigItem(
+		keyName = "unfocusedLimitFps",
+		name = "Limit FPS When Unfocused",
+		description = "Stay at or under the target frames per second when unfocused",
+		position = 1
+	)
+	default boolean unfocusedLimitFps()
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsConfig.java]: unfocusedLimitFps and alwaysLimitFps should be probably 1 option switched between with enum (e.g Limit: ALWAYS, UNFOCUSED, OFF)

**Comment by LeviSchuck** [comment]
Updated @deathbeam

File: runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsLimitMode.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
+package net.runelite.client.plugins.fps;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsLimitMode.java]: You forgot license header

Comment (LeviSchuck) [reply runelite-client/src/main/java/net/runelite/client/plugins/fps/FpsLimitMode.java]: it do now.
